Fix unique_id for components not overriding default.

frolswe opened the pull request:
Home Assistant gets confused if unique_id is not globally unique. E.g.
if two different devices both use unique_id="ESPlightlight". This
commit fixes this by adding the MAC address to the id, as is already
done by some components.

Unfortunately this will change the id for some component types (notably
light, and switch), which will make HA add them as new entities. A
workaround for this is to remove the devices from the HA MQTT
integration, and then reboot the devices and they will automatically be
discovered again. Now the HA entity_id can be changed to the desired
name.

Some examples from other manufacturers:
- ZigBee 2 MQTT
  unique_id = "0x14b457fffe71cbbf_click_zigbee2mqtt"
- Tasmota
  uniq_id = "5A5222_LI_1"
  uniq_id = "55C951_ENERGY_Total"
- Shelly EM3
  unique_id = "C82B96116645/emeter/0/current"

Signed-off-by: Fredrik Olofsson <frol@hupp.se>

frolswe committed:
Fix unique_id for components not overriding default.

Home Assistant gets confused if unique_id is not globally unique. E.g.
if two different devices both use unique_id="ESPlightlight". This
commit fixes this by adding the MAC address to the id, as is already
done by some components.

Unfortunately this will change the id for some component types (notably
light, and switch), which will make HA add them as new entities. A
workaround for this is to remove the devices from the HA MQTT
integration, and then reboot the devices and they will automatically be
discovered again. Now the HA entity_id can be changed to the desired
name.

To enable this new unique id, set the mqtt.use_new_unique_id
configuration to true.

Some examples from other manufacturers:
- ZigBee 2 MQTT
  unique_id = "0x14b457fffe71cbbf_click_zigbee2mqtt"
- Tasmota
  uniq_id = "5A5222_LI_1"
  uniq_id = "55C951_ENERGY_Total"
- Shelly EM3
  unique_id = "C82B96116645/emeter/0/current"

Signed-off-by: Fredrik Olofsson <frol@hupp.se>

OttoWinter commented:
Yeah the unique ID currently is not good for MQTT. In native API we use the device name for building the unique id (device name is required to be unique in the network anyway).

As you said, this would invalidate all previous unique IDs, which is not good. I'm not sure the cost of breaking everybody's HA installs (because the entity ID will now have a `_2` suffix) is worth it. At least I've heard some pretty bad UX from unique ID changes with ZHA.

Ideally, HA should add some mechanism to update unique IDs of existing entities. Until that exists, could be maybe hide this behind an opt-in config option, like 

```yaml
mqtt:
  use_new_unique_id: yes
```

Thanks!

frolswe commented:
I rebased the PR and added the use_new_unique_id configuration. Though, I do not find my way around the esphome codebase very well, so I might not have added the configuration in the best way..

glmnet commented on esphome/components/mqtt/__init__.py:
```diff
@@ -165,6 +166,8 @@ def to_code(config):
     cg.add(var.set_password(config[CONF_PASSWORD]))
     if CONF_CLIENT_ID in config:
         cg.add(var.set_client_id(config[CONF_CLIENT_ID]))
+    if CONF_USE_NEW_UNIQUE_ID in config:
```
remove this line, ``CONF_USE_NEW_UNIQUE_ID`` has a default value so this is always true, also you're not initializing the field in the c++ class so you must call the setter

glmnet commented on esphome/config_validation.py:
```diff
@@ -1276,6 +1277,7 @@ def validate(value):
     Optional(CONF_AVAILABILITY): All(requires_component('mqtt'),
                                      Any(None, MQTT_COMPONENT_AVAILABILITY_SCHEMA)),
     Optional(CONF_INTERNAL): boolean,
+    Optional(CONF_USE_NEW_UNIQUE_ID): boolean,
```
What's this one for? Is there some magic behind  the scenes which overrides the settings for a single sensor/binary_sensor, etc?

glmnet reviewed (changes requested):
Please add docs entry meanwhile too, this helps us review code better as we read docs to understand your intentions in code. Thanks
